server/src-lib/Data/Time/Clock/Units.hs

```diff
@@ -148,24 +371,70 @@ instance (KnownNat picosPerUnit) => RealFrac (TimeUnit picosPerUnit) where
   floor = floor . toRational
 
 -- we can ignore unit:
-instance Hashable (TimeUnit a) where
-  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $ 
+instance Hashable (TimeUnit a 'Absolute) where
+  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $
     (realToFrac :: DiffTime -> Double) dt
 
 
+-- instances for Calendar
+-- NOTE: secondsToNominalDiffTime :: Pico -> NominalDiffTime
+-- http://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Clock.html
+-- If a Pico value is supplied to the function then the output is also in Pico
+-- and when the Pico value is converted into seconds(divide by 10^12)
+-- then the output is in Seconds
+-- So, the function `toFixedPico` converts Pico into Seconds
+instance (KnownNat picosPerUnit) => Num (TimeUnit picosPerUnit 'Calendar) where
+  TimeUnit a + TimeUnit b = TimeUnit $ a + b
+  TimeUnit a - TimeUnit b = TimeUnit $ a - b
+  TimeUnit a * TimeUnit b = TimeUnit . secondsToNominalDiffTime . toFixedPico $
+    (toRational $ nominalDiffTimeToSeconds a * nominalDiffTimeToSeconds b) `div'` (natNum @picosPerUnit)
+  negate (TimeUnit a) = TimeUnit $ negate a
+  abs (TimeUnit a) = TimeUnit $ abs a
+  signum (TimeUnit a) = TimeUnit $ signum a
+  fromInteger a = TimeUnit . secondsToNominalDiffTime . toFixedPico $ a * natNum @picosPerUnit
+
+instance (KnownNat picosPerUnit) => Read (TimeUnit picosPerUnit 'Calendar) where
+  readsPrec _ = map (first fromRational) . readFloat
+
+instance (KnownNat picosPerUnit) => Fractional (TimeUnit picosPerUnit 'Calendar) where
+  TimeUnit a / TimeUnit b = TimeUnit . secondsToNominalDiffTime . toFixedPico $ fromInteger $
+    ((toRational $ nominalDiffTimeToSeconds a) * natNum @picosPerUnit) `div'` (toRational $ nominalDiffTimeToSeconds b)
+  fromRational a = TimeUnit . secondsToNominalDiffTime . toFixedPico $ round (a * natNum @picosPerUnit)
+
+instance (KnownNat picosPerUnit) => Real (TimeUnit picosPerUnit 'Calendar) where
+  toRational (TimeUnit a) = toRational (nominalDiffTimeToSeconds a) / natNum @picosPerUnit
+
+instance (KnownNat picosPerUnit) => RealFrac (TimeUnit picosPerUnit 'Calendar) where
+  properFraction a = (i, a - fromIntegral i)
+    where i = truncate a
+  truncate = truncate . toRational
+  round = round . toRational
+  ceiling = ceiling . toRational
+  floor = floor . toRational
+
+-- we can ignore unit:
+instance Hashable (TimeUnit a 'Calendar) where
+  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $
+    (realToFrac :: NominalDiffTime -> Double) dt
+
 -- | Duration types isomorphic to 'DiffTime', powering 'fromUnits'.
-class Duration d where
+class Duration' d where
```

Comment:
I’d rename this class to `AsDuration` rather than `Duration'`. It’s the naming convention `lens` uses for classes like this.